File: src/com/suushiemaniac/cubing/bld/filter/BooleanCondition.java
package com.suushiemaniac.cubing.bld.filter;

public class BooleanCondition {
    public static BooleanCondition YES() {
        return new BooleanCondition(true, true);
    }

    public static BooleanCondition NO() {
        return new BooleanCondition(false, true);
    }

    public static BooleanCondition UNIMPORTANT() {
        return new BooleanCondition(false, false);
    }

    private boolean truthValue, isImportant;

    private BooleanCondition(boolean truthValue, boolean isImportant) {
        this.truthValue = truthValue;
        this.isImportant = isImportant;
    }

    public void setImportant(boolean important) {
        this.isImportant = important;
    }

    public boolean isImportant() {
        return this.isImportant;
    }

    public boolean getValue() {
        return this.truthValue;
    }

    public void setValue(boolean newValue) {
        this.truthValue = newValue;
    }

    public boolean getPositive() {
        return !this.isImportant || this.truthValue;
    }

    public boolean getNegative() {
        return this.isImportant && this.truthValue;
    }

    public boolean evaluatePositive(boolean compareTo) {
        return !this.isImportant || this.truthValue == compareTo;
    }

    public boolean evaluateNegative(boolean compareTo) {
        return this.isImportant && this.truthValue == compareTo;
    }

    @Override
    public String toString() {
        return this.isImportant ? String.valueOf(this.truthValue) : "unimportant";
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/bld/filter/condition/BooleanCondition.kt
package com.suushiemaniac.cubing.bld.filter.condition

data class BooleanCondition(private var value: Boolean, val isImportant: Boolean) : BaseConditional<Boolean> {
    val positive: Boolean
        get() = !this.isImportant || this.value

    val negative: Boolean
        get() = this.isImportant && this.value

    fun evaluatePositive(compareTo: Boolean) = !this.isImportant || this.value == compareTo
    fun evaluateNegative(compareTo: Boolean) = this.isImportant && this.value == compareTo

    override fun evaluate(compareTo: Boolean) = this.evaluatePositive(compareTo)

    fun define(value: Boolean) {
        this.value = value
    }

    companion object {
        fun YES() = BooleanCondition(value = true, isImportant = true)

        fun NO(): BooleanCondition = BooleanCondition(value = false, isImportant = true)

        fun UNIMPORTANT() = BooleanCondition(value = false, isImportant = false)
        fun MAYBE() = UNIMPORTANT()

        fun STRICT(value: Boolean, isStrict: Boolean = false) = if (value) if (isStrict) YES() else MAYBE() else NO()
    }
}
